### DISCUSSION TITLE
TailwindCSS not installing correctly as (production) dependency

### DISCUSSION THREAD

**Discussion by wstrother** [vercel/community#4004] [Help]
### Page to Investigate

https://streaming-app-tau.vercel.app/api/css?css=bg-red-500

### Steps to Reproduce

The endpoint listed above invokes postcss / tailwind programmatically but produces a 500 error with the following stack trace:

```
Error: ENOENT: no such file or directory, open '/var/task/node_modules/tailwindcss/lib/css/preflight.css'
    at Object.openSync (node:fs:603:3)
    at Object.readFileSync (node:fs:471:35)
    at preflight (/var/task/node_modules/tailwindcss/lib/corePlugins.js:491:66)
    at registerPlugins (/var/task/node_modules/tailwindcss/lib/lib/setupContextUtils.js:794:61)
    at createContext (/var/task/node_modules/tailwindcss/lib/lib/setupContextUtils.js:1196:5)
    at getContext (/var/task/node_modules/tailwindcss/lib/lib/setupContextUtils.js:1253:19)
    at /var/task/node_modules/tailwindcss/lib/lib/setupTrackingContext.js:118:81
    at /var/task/node_modules/tailwindcss/lib/processTailwindFeatures.js:48:11
    at plugins (/var/task/node_modules/tailwindcss/lib/plugin.js:38:69)
    at LazyResult.runOnRoot (/var/task/node_modules/postcss/lib/lazy-result.js:357:16) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/var/task/node_modules/tailwindcss/lib/css/preflight.css'
}
```

It appears the file `node_modules\tailwindcss\lib\css\preflight.css` is not being installed with the rest of TailwindCSS.

Tailwind is also invoked by Vite at build time and this apparently works successfully because all of the static CSS is generated the same as in my local production preview, it's only when this endpoint calls TailwindCSS programmatically by this api endpoint that the error occurs.

**Comment by amyegan** [comment]
Hey @wstrother! Do you have a $[minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to help with debugging?

**Reply by wstrother** [reply]
Yes, you can clone https://github.com/wstrother/streaming-app/tree/dev/mrp-vercel-tailwind which just implements the single API route and uses the sveltekit vercel adapter with only the Tailwind/Postcss dependencies.

**Reply by amyegan** [reply]
Can you help me understand why you want to configure Tailwind in a server route? It works without error to style the page on your example branch when I use the Tailwind-recommended instructions: https://tailwindcss.com/docs/guides/sveltekit

**Reply by wstrother** [reply]
Because the entire purpose of the app is to allow for customizable stream layouts to be targeted by OBS. Tailwind as it's normally configured works by statically analyzing a list of files with regex to search for the relevant utility classes at build time. The classes that get applied to a user's layout aren't there at build time, they are either retrieved from the database or entered as user input on the client side.

**Reply by amyegan** [reply]
Hey Wyatt. I was trying to understand the end goal so I could help you come up with an appropriate solution. I've never seen anyone do this before, so I'm not sure if it's possible. I'll need to try a few things.

It may be worth reaching out in [Tailwind CSS Discussions](https://github.com/tailwindlabs/tailwindcss/discussions), [SvelteKit Discussions](https://github.com/sveltejs/kit/discussions), or [Svelte Discord](https://svelte.dev/chat) to get the opinions of the people who work most closely with those tools.

**Reply by wstrother** [reply]
Amy, reaching out to those communities would be irrelevant because your platform is what's causing the problem. Running the production build / preview process reveals that it is indeed possible to invoke Tailwind programmatically this way. Tailwind is just a plugin for PostCSS, which is a library that generates an AST for CSS that plugins like Tailwind can use to extend new functionalities to CSS. Ultimately it is literally just a function that takes a string and turns it into another string. The fact that you're "not sure" if this is possible doesn't reflect well on your platform.

A quick Google search reveals dozens of similar questions in the Tailwind community, as well as people giving the correct solution and others reporting happily that it works. [Here is a repository](https://github.com/bradlc/tailwindcss-as-a-service/) that literally just implements a server that generates tailwind css classes based on a request.

I'm sorry if I seem frustrated, but I've given you a stack trace that makes it quite obvious what the problem is: your platform is apparently tree shaking a CSS file out of a node module that is imported by an API route. Something has obviously gone wrong. Obviously I understand the need to minimize bundle sizes it general, but when the process you use to do this is removing files that are used at runtime and causing an error, something has gone wrong. 

Apparently I needed to be more explicit that Vercel's deployment was causing the issue, so I apologize for not making that clear. I would've hoped it would be obvious based on the context that I am asking on the Vercel community page specifically for support about your platform, that I'm not in fact asking whether it is possible to implement this functionality that I've already implemented. Especially when I provided you with a minimum reproducible example that you can clone and literally type 'npm run build | npm run preview' and in 5 minutes conclusively confirm that the API route works fine when being built locally.

**Reply by cowboycodr** [reply]
Did you ever find a solution for this?

I'm getting the same error.
